tools/clang/test/HLSLFileCheck/rewriter/remove-unused-globals.hlsl

```diff
@@ -0,0 +1,65 @@
+// RUN: %dxr -E main -remove-unused-globals %s | FileCheck %s
+
+// CHECK: cbuffer TestCBuffer
+cbuffer TestCBuffer
+{
+// CHECK: int TestInt;
+  int TestInt;
+// UnusedFloat is not removed if inside a cbuffer declaration with a used global
+// CHECK: float UnusedFloat;
+  float UnusedFloat;
+}
+// CHECK: }
+
+// Unused cbuffers are not removed at this time
+// CHECK: cbuffer UnusedCBuffer
+cbuffer UnusedCBuffer
+{
+// CHECK: float FloatInUnusedCBuffer;
+  float FloatInUnusedCBuffer;
+}
+// CHECK: }
+
+// CHECK-NOT: RandomGlobal
+float4 RandomGlobal[2];
+
+// CHECK: LookupTable
+float4 LookupTable[1];
+
+// CHECK-NOT: UnusedBool
+bool UnusedBool;
+
+// CHECK: TestBool
+bool TestBool;
+
+// CHECK-NOT: UnusedTex
+Texture2D UnusedTex;
+
+// CHECK: TestTex
+Texture2D TestTex;
+
+// CHECK-NOT: UnusedSampler
+SamplerState UnusedSampler;
+
+// CHECK: TestSampler
+SamplerState TestSampler;
+
+// CHECK: void main(
+void main(in float4 SvPosition : SV_Position,
+          out float4 OutTarget0 : SV_Target0)
+{
+  float3 Color = 0;
+  uint Index = 0;
+
+  if (TestBool)
+  {
+    Index = TestInt;
+  }
+
+  float2 UV = LookupTable[Index].xy + LookupTable[Index].zw;
+  float4 Value = TestTex.SampleLevel(TestSampler, UV, 0);
+
+  Color += Value.rgb + (1 - Value.a);
+
+  OutTarget0 = float4 (Color, 1.0f);
```

Comment:
What happens with a global that is used, but doesn't contribute to the output variable? Would that be expected to be removed? Either way, it might be an interesting additional test.

Reply:
I added a referenced, but unused value to the test to verify that the value is kept.  Note: It would be far more difficult for the rewriter to remove such values, since it's rewriting the HLSL, and it would have to replace the actual references to these values with something semantically compatible as well as the global values themselves.  It would likely have to replace all references to constant values with dummy immediate values with the appropriate type.  But some types would have additional complications, such as non-constant globals, resources, structured types (where we want to remove the structure when it has no other active users), and probably more.  That would be a much larger job.